feedback/invalid_token.go

```diff
@@ -0,0 +1,224 @@
+	"github.com/topfreegames/pusher/extensions"
+	"github.com/topfreegames/pusher/interfaces"
+	"github.com/topfreegames/pusher/util"
+)
+
+// InvalidToken represents a token with the necessary information to be deleted
+type InvalidToken struct {
+	Token    string
+	Game     string
+	Platform string
+}
+
+// InvalidTokenHandler takes the InvalidTokens from the InChannel and put them in a buffer.
+// When the buffer is full or after a timeout, it is flushed, triggering the deletion
+// of the tokens from the database
+type InvalidTokenHandler struct {
+	Logger *log.Logger
+	Config *viper.Viper
+	Client *extensions.PGClient
+
+	FlushTicker *time.Ticker
+	InChan      *chan *InvalidToken
+	Buffer      []*InvalidToken
+	bufferSize  int
+	run         bool
+	stopChan    chan bool
+}
+
+// NewInvalidTokenHandler returns a new InvalidTokenHandler instance
+func NewInvalidTokenHandler(
+	logger *log.Logger, cfg *viper.Viper,
+	inChan *chan *InvalidToken,
+	dbOrNil ...interfaces.DB,
+
+) (*InvalidTokenHandler, error) {
+	h := &InvalidTokenHandler{
+		Logger:   logger,
+		Config:   cfg,
+		InChan:   inChan,
+		stopChan: make(chan bool),
+	}
+
+	var db interfaces.DB
+	if len(dbOrNil) == 1 {
+		db = dbOrNil[0]
+	}
+
+	err := h.configure(db)
+	if err != nil {
+		return nil, err
+	}
+
+	return h, nil
+}
+
+func (i *InvalidTokenHandler) loadConfigurationDefaults() {
+	i.Config.SetDefault("feedbackListeners.invalidToken.flush.time.ms", 5000)
+	i.Config.SetDefault("feedbackListeners.invalidToken.buffer.size", 1000)
+}
+
+func (i *InvalidTokenHandler) configure(db interfaces.DB) error {
+	l := i.Logger.WithField("method", "configure")
+	i.loadConfigurationDefaults()
+
+	flushTime := time.Duration(i.Config.GetInt("feedbackListeners.invalidToken.flush.time.ms")) * time.Millisecond
+	i.bufferSize = i.Config.GetInt("feedbackListeners.invalidToken.buffer.size")
+
+	i.FlushTicker = time.NewTicker(flushTime)
+	i.Buffer = make([]*InvalidToken, 0, i.bufferSize)
+
+	var err error
+	i.Client, err = extensions.NewPGClient("feedbackListeners.invalidToken.pg", i.Config, db)
+	if err != nil {
+		l.WithError(err).Error("failed to configure psql database")
+		return err
+	}
+
+	l.Info("psql database configured")
+	return nil
+}
+
+// Start starts to process the InvalidTokens from the intake channel
+func (i *InvalidTokenHandler) Start() {
+	l := i.Logger.WithField(
+		"method", "start",
+	)
+	l.Info("starting invalid token handler")
+
+	i.run = true
+	go i.processMessages()
+}
+
+// Stop stops the Handler from consuming messages from the intake channel
+func (i *InvalidTokenHandler) Stop() {
+	i.run = false
+	close(i.stopChan)
+}
+
+func (i *InvalidTokenHandler) processMessages() {
+	l := i.Logger.WithFields(log.Fields{
+		"method": "processMessages",
+	})
+
+	for i.run {
+		select {
+		case tk, ok := <-*i.InChan:
+			if ok {
+				i.Buffer = append(i.Buffer, tk)
+
+				if len(i.Buffer) >= i.bufferSize {
+					l.Debug("buffer is full")
+					go i.deleteTokens(i.Buffer)
```

Comment:
why are you running deleteTokens in a go routine? shouldn't it block the loop? what happens if you run the next line before fetching `i.Buffer`? I'm sure we'll have concurrency issues here!

Reply:
I run it in a different routine to avoid blocking the handler from processing the intake channel.
Since we pass a slice, the pointer to the actual array that stores the data is copied and we shouldn't lose the reference to it inside the new go routine. 

I've done a code snippet to test it:
https://play.golang.org/p/nwvL2_idURD

Reply:
if by some reason the deleteTokens procedure is too slow, wouldn't this mess with the whole process because of innumerous goroutines being created?

remember that we have no rush on deleting this tokens, it would be ok even if we deleted them synchronously, it's not like the tokens are going somewhere else (they'll stay in kafka 'till we process them, at least for "retention period")

if you want to do it asynchronously, at least put some limit in the number of running goroutines, the approach I prefer is using a channel for receiving the "i.Buffer" and having a pre-established number of goroutines handling them, that way go would do the blocking job for us.

Reply:
Got it. I'll make the whole process synchronous them, including the broker. We can release this first version and see what is the pipeline bottleneck. Then we can tackle the problem by implementing a pool of worker to perform the deletion if it's necessary. 

In a next PR I'll send metrics reporting the channels size to help us getting some insights.

Reply:
ok!